### ISSUE TITLE
Writing safe code with Kotlin

### ISSUE THREAD

**Issue by Haozhe321** [nus-cs3281/2018#26]
**WIIFY:**
Prevent your android apps from crashing less(~30%) with Kotlin. Less complains from customers and more praises from developers for writing concise yet informative code. 

**Key Points:** 
1. NullPointerException accounts for 30% of all Exceptions in Android
2. Kotlin's type system can help to eliminate NullPointerException 
3. Error is detected at compile time instead of runtime. 

**Call to action:**
1. Switch to Kotlin for safer type system when developing your next android app. 
2. Use non-nullable type if you don't expect to have a null for your variable. 

Slides: 
[lightning talk 1.pptx](https://github.com/nus-cs3281/2018/files/1716685/lightning.talk.1.pptx)

**Comment by Haozhe321** [comment]
@LiHaoTan asked a great question during Q&A: why let null propagate in the program instead of throwing a NullPointerException and terminating the program? 

My response is following: 
Kotlin's type system is aimed at eliminating NullPointerException because every time an app crashes, money is lost. That said, the Elvis operator (?:) can be used to **prevent the propagation of null through the code.** Having the generic `NullPointerException` is difficult to debug, and the Elvis operator can be used to provide a descriptive message, e.g. 
`val name = node.getName() ?: throw IllegalArgumentException("name expected")` 
or we can use the `requireNotNull` method and providing a message as well, e.g.
`uploadPhoto(requireNotNull(intent.getStringExtra("PHOTO_URL"), { "Activity parameter 'PHOTO_URL' is missing" }))`

In the above examples, the `null` will not propagate through the program, and we can easily debug our program. So even if a `null` is returned in the case of a safe call operator(?.), Kotlin has additional layers of safety checks(by verifying the returned value and dealing with it) instead of crashing your app.  In many instances, we don't actually need to declare a type as nullable(non-nullable types are safer), but Kotlin allows nullable type because Kotlin is meant to be interoperable with Java, so `null` reference has to be supported. There are many discussions of completely removing `null` references(including the inventor of null) because it does more harm than good.

In summary: 
1. Philosophy of safe call operator: returning a `null` is better than crashing the app, it's safely handled and programmers can deal with the `null` value without app crash. 
2. There are ways to stop the propagation of `null` through the program if a `null` is returned; the Elvis operator can be used to assign a default value in the case of a `null`, or provide a descriptive message of the unexpected value. 
3. Avoid using nullable types
